parse.py

```diff
@@ -138,55 +137,94 @@ def query_pageids(pageids):
+        except:
+            traceback.print_exc()
+            self.exception_count += 1
+            if self.exception_count > MAX_EXCEPTIONS_PER_SUBPROCESS:
+                print('Too many exceptions, quitting!')
+                raise
+    return wrapper
+
+@with_max_exceptions
+def work(model, pageids):
+    rows = []
+    results = query_pageids(pageids)
+    for revid, title, wikitext in results:
+        start_len = len(rows)
+        wikicode = mwparserfromhell.parse(wikitext)
+        sentences, paragraphs = extract(wikicode)
+
+        for text, pidx, section in sentences:
+            row = [text, paragraphs[pidx], section, revid]
+            rows.append(row)
+
+        X, sections = run_citation_needed(sentences)
+        pred = model.predict([X, sections])
+        for i, score in enumerate(pred):
+            rows[start_len+i].append(score[1])
 
     def insert(cursor, r):
         cursor.execute('''
             INSERT INTO statements (statement, context, section, rev_id, score)
             VALUES(%s, %s, %s, %s, %s)
             ''', r)
     db = cddb.init_scratch_db()
+    for r in rows:
+        try:
+            db.execute_with_retry(insert, r)
+        except:
+            # This may be caused by the size
+            # of a sentence or paragraph that
+            # exceed the maximum limit
+            continue
 
-    batch_size = 32
-    for i in range(0, len(pageids_list), batch_size):
-        rows = []
-        results = query_pageids(pageids_list[i: i + batch_size])
-        for revid, title, wikitext in results:
-            start_len = len(rows)
-            wikicode = mwparserfromhell.parse(wikitext)
-            sentences, paragraphs = extract(wikicode)
-
-            for text, pidx, section in sentences:
-                row = [text, paragraphs[pidx], section, revid]
-                rows.append(row)
-
-            pred = run_citation_needed(sentences, model, vocab_w2v, section_dict)
-            for i, score in enumerate(pred):
-                rows[start_len+i].append(score[1])
+def parse(pageids, timeout):
+    from keras_model import KerasManager
+    manager = KerasManager()
+    manager.start()
+    model = manager.KerasModel()
+    model.initialize()
 
-            page_processed += 1
+    pool = multiprocessing.Pool(
+        processes = multiprocessing.cpu_count(), 
+        initializer = initializer)
 
-        for r in rows:
-            try:
-                db.execute_with_retry(insert, r)
-            except:
-                # This may be caused by the size
-                # of a sentence or paragraph that
-                # exceed the maximum limit
-                continue
-
-    print('page processed: ', page_processed)
-    return 0
+    tasks = []
+    batch_size = 32
+    pageids_list = list(pageids)
+    for i in range(0, len(pageids_list), batch_size):
+        tasks.append(pageids_list[i:i+batch_size])
+
+    result = pool.map_async(functools.partial(work, model), tasks)
```

Comment:
I think you could just make the model a global variable instead of passing it in with `functools.partial`, no?

Reply:
Yes, you are right! I thought if I make the model a global variable, the model would be copied to each child process and that's not what we want.. because we want to access a shared model between all processes. Did I misunderstand something?

Reply:
multiprocessing managers is allow you to create an object in a separate process and manipulate it via "proxy" objects in your process -- then, multiprocessing figures out how to communicate with the actual object, sending method calls and return values back and forth.

As I understand it, the model is created and loaded in a separate process, and your `model` variable is just a proxy for it, which should be totally fine to copy and pass around to other processes.

Try running this gist and looking at the output: https://gist.github.com/eggpi/c173f2bd541dc931f51e515e9a8b4e64, it shows in which processes the different things happen :)

Reply:
I see what you mean!! That's cool!!
It's very clear through your example code. Thank you :D